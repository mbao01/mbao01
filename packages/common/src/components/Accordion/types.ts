import * as AccordionPrimitive from "@radix-ui/react-accordion";

export type AccordionProps = React.ComponentPropsWithoutRef<typeof AccordionPrimitive.Root>;

export type AccordionContentProps = React.ComponentPropsWithoutRef<
  typeof AccordionPrimitive.Content
>;

export type AccordionItemProps = React.ComponentPropsWithoutRef<typeof AccordionPrimitive.Item>;

export type AccordionTriggerProps = React.ComponentPropsWithoutRef<
  typeof AccordionPrimitive.Trigger
>;
