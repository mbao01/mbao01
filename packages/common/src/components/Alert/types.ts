import { type VariantProps } from "../../libs";
import { getAlertClasses } from "./constants";

export type AlertProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof getAlertClasses>;

export type AlertTitleProps = React.HTMLAttributes<HTMLHeadingElement>;

export type AlertDescriptionProps = React.HTMLAttributes<HTMLDivElement>;
