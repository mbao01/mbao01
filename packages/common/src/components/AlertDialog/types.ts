import * as AlertDialogPrimitive from "@radix-ui/react-alert-dialog";
import { type VariantProps } from "../../libs";
import { getButtonClasses } from "../Button/constants";
import {
  getDialogContentClasses,
  getDialogDescriptionClasses,
  getDialogFooterClasses,
  getDialogHeaderClasses,
  getDialogOverlayClasses,
  getDialogTitleClasses,
} from "../Dialog/constants";

export type AlertDialogProps = React.ComponentPropsWithoutRef<typeof AlertDialogPrimitive.Root>;

export type AlertDialogOverlayProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Overlay
> &
  VariantProps<typeof getDialogOverlayClasses>;

export type AlertDialogContentProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Content
> &
  VariantProps<typeof getDialogContentClasses>;

export type AlertDialogHeaderProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof getDialogHeaderClasses>;

export type AlertDialogFooterProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof getDialogFooterClasses>;

export type AlertDialogTitleProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Title
> &
  VariantProps<typeof getDialogTitleClasses>;

export type AlertDialogDescriptionProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Description
> &
  VariantProps<typeof getDialogDescriptionClasses>;

export type AlertDialogActionProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Cancel
> &
  VariantProps<typeof getButtonClasses>;

export type AlertDialogCancelProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Cancel
> &
  VariantProps<typeof getButtonClasses>;
