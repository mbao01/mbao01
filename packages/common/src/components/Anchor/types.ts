import { VariantProps } from "../../libs";
import { getAnchorClasses } from "./constant";

export type AnchorProps = React.AnchorHTMLAttributes<HTMLAnchorElement> & {
  isExternal?: boolean;
} & VariantProps<typeof getAnchorClasses>;
