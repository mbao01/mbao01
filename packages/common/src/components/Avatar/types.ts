import * as AvatarPrimitive from "@radix-ui/react-avatar";
import { type VariantProps } from "../../libs";
import {
  getAvatarClasses,
  getAvatarFallbackClasses,
  getAvatarImageClasses,
} from "./constants";

export type AvatarProps = Omit<
  React.ComponentPropsWithoutRef<typeof AvatarPrimitive.Root>,
  "asChild" | "size"
> &
  VariantProps<typeof getAvatarClasses>;

export type AvatarImageProps = Omit<
  React.ComponentPropsWithoutRef<typeof AvatarPrimitive.Image>,
  "asChild"
> &
  VariantProps<typeof getAvatarImageClasses>;

export type AvatarFallbackProps = Omit<
  React.ComponentPropsWithoutRef<typeof AvatarPrimitive.Fallback>,
  "asChild"
> &
  VariantProps<typeof getAvatarFallbackClasses>;
