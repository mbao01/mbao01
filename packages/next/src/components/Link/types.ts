import type { LinkProps as OriginalLinkProps } from "next/dist/client/link.js";
import { AnchorHTMLAttributes, DetailedHTMLProps } from "react";
import { type UrlObject } from "url";
import { type VariantProps } from "@mbao01/common/libs";
import { getLinkClasses } from "./constant";

export type LinkProps<T> = Omit<
  Omit<
    DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>, HTMLAnchorElement>,
    keyof OriginalLinkProps
  > &
    OriginalLinkProps,
  "href"
> & { isExternal?: boolean; hardNavigate?: boolean } & (
    | {
        href: __next_route_internal_types__.RouteImpl<T> | UrlObject;
        isInternal?: true;
      }
    | {
        href: string;
        isInternal?: false;
      }
  ) &
  VariantProps<typeof getLinkClasses>;
