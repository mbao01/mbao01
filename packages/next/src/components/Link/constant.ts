import { cva } from "@mbao01/common/libs";

export const getLinkClasses = cva("link transition-all", {
  variants: {
    variant: {
      accent: "link-accent",
      default: "link-default",
      error: "link-error",
      info: "link-info",
      link: "link-link",
      neutral: "link-neutral",
      primary: "link-primary",
      secondary: "link-secondary",
      success: "link-success",
      warning: "link-warning",
    },
    hover: {
      true: "link-hover",
    },
    underline: {
      false: "no-underline",
    },
  },
});
