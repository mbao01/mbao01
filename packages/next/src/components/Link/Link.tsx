import NextLink from "next/link";
import { ExternalLinkIcon } from "@radix-ui/react-icons";
import { Anchor } from "@mbao01/common";
import { cn } from "@mbao01/common/utilities";
import { type LinkProps } from "./types";
import { getLinkClasses } from "./constant";

export const Link = <T,>({
  href,
  hover,
  target,
  variant,
  children,
  className,
  underline = false,
  isExternal = false,
  isInternal = true,
  ...props
}: LinkProps<T>) => {
  if (isInternal === false) {
    return (
      <Anchor
        href={href as string}
        target={target}
        hover={hover}
        variant={variant}
        className={className}
        underline={underline}
        isExternal={isExternal}
        {...props}
      >
        {children}
      </Anchor>
    );
  }

  return (
    <NextLink
      href={href as Omit<LinkProps<T>["href"], string>}
      target={target}
      className={cn(getLinkClasses({ hover, variant, underline }), className)}
      {...props}
      {...(isExternal && { rel: "noopener noreferrer" })}
    >
      {children}
      {isExternal ? (
        <ExternalLinkIcon name="external" className="ml-[2px] inline" />
      ) : null}
    </NextLink>
  );
};
