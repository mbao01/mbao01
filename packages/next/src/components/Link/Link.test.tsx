import { render, screen } from "@testing-library/react";

import { Link } from "./Link";
import { LinkProps } from "./types";
import { MemoryRouter, MemoryRouterProps } from "react-router-dom";

describe("Link", () => {
  const renderLink = (
    props: LinkProps<string>,
    memoryRouterProps?: MemoryRouterProps
  ) => {
    return render(
      <MemoryRouter {...memoryRouterProps}>
        <Link {...props} />
      </MemoryRouter>
    );
  };

  it("renders a link with correct label", () => {
    const { asFragment } = renderLink({
      hover: false,
      children: "Go Home 🏡",
      href: { pathname: "/home", search: "hello=world", hash: "great" },
    });

    const anchorEl = screen.getByRole("link", { name: "Go Home 🏡" });

    expect(anchorEl).toBeInTheDocument();
    expect(anchorEl).toHaveAttribute("href", "/home?hello=world#great");
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders an external link", () => {
    const { asFragment } = renderLink({
      href: "https://example.com",
      target: "_blank",
      children: "Google.com",
    });

    const anchorEl = screen.getByRole("link", { name: "Google.com" });

    expect(anchorEl).toBeInTheDocument();
    expect(anchorEl).toHaveAttribute("href", "https://example.com");
    expect(asFragment()).toMatchSnapshot();
  });

  it.each([
    "accent",
    "default",
    "error",
    "info",
    "link",
    "neutral",
    "primary",
    "secondary",
    "success",
    "warning",
  ] as const)("has %s variant", (variant) => {
    const label = `${variant} link`;
    const { asFragment } = renderLink({
      hover: true,
      children: label,
      href: `/variant/${variant}` as LinkProps<string>["href"],
    });

    const anchorEl = screen.getByRole("link", { name: label });
    expect(anchorEl).toBeInTheDocument();
    expect(anchorEl).toHaveAttribute("href", `/variant/${variant}`);
    expect(asFragment()).toMatchSnapshot();
  });
});
