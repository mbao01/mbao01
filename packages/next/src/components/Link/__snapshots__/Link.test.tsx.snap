// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Link > has accent variant 1`] = `
<DocumentFragment>
  <a
    class="link transition-all link-hover no-underline"
    href="/variant/accent"
  >
    accent link
  </a>
</DocumentFragment>
`;

exports[`Link > has default variant 1`] = `
<DocumentFragment>
  <a
    class="link transition-all link-hover no-underline"
    href="/variant/default"
  >
    default link
  </a>
</DocumentFragment>
`;

exports[`Link > has error variant 1`] = `
<DocumentFragment>
  <a
    class="link transition-all link-hover no-underline"
    href="/variant/error"
  >
    error link
  </a>
</DocumentFragment>
`;

exports[`Link > has info variant 1`] = `
<DocumentFragment>
  <a
    class="link transition-all link-hover no-underline"
    href="/variant/info"
  >
    info link
  </a>
</DocumentFragment>
`;

exports[`Link > has link variant 1`] = `
<DocumentFragment>
  <a
    class="link transition-all link-hover no-underline"
    href="/variant/link"
  >
    link link
  </a>
</DocumentFragment>
`;

exports[`Link > has neutral variant 1`] = `
<DocumentFragment>
  <a
    class="link transition-all link-hover no-underline"
    href="/variant/neutral"
  >
    neutral link
  </a>
</DocumentFragment>
`;

exports[`Link > has primary variant 1`] = `
<DocumentFragment>
  <a
    class="link transition-all link-hover no-underline"
    href="/variant/primary"
  >
    primary link
  </a>
</DocumentFragment>
`;

exports[`Link > has secondary variant 1`] = `
<DocumentFragment>
  <a
    class="link transition-all link-hover no-underline"
    href="/variant/secondary"
  >
    secondary link
  </a>
</DocumentFragment>
`;

exports[`Link > has success variant 1`] = `
<DocumentFragment>
  <a
    class="link transition-all link-hover no-underline"
    href="/variant/success"
  >
    success link
  </a>
</DocumentFragment>
`;

exports[`Link > has warning variant 1`] = `
<DocumentFragment>
  <a
    class="link transition-all link-hover no-underline"
    href="/variant/warning"
  >
    warning link
  </a>
</DocumentFragment>
`;

exports[`Link > renders a link with correct label 1`] = `
<DocumentFragment>
  <a
    class="link transition-all no-underline"
    href="/home?hello=world#great"
  >
    Go Home 🏡
  </a>
</DocumentFragment>
`;

exports[`Link > renders a non internal link 1`] = `
<DocumentFragment>
  <a
    class="link transition-all no-underline"
    href="https://example.com"
  >
    Google.com
  </a>
</DocumentFragment>
`;

exports[`Link > renders an external link 1`] = `
<DocumentFragment>
  <a
    class="link transition-all no-underline"
    href="https://example.com"
    rel="noopener noreferrer"
  >
    Google.com
    <svg
      aria-hidden="true"
      class="lucide lucide-external-link ml-[2px] w-4 h-4 inline"
      fill="none"
      height="24"
      name="external"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M15 3h6v6"
      />
      <path
        d="M10 14 21 3"
      />
      <path
        d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
      />
    </svg>
  </a>
</DocumentFragment>
`;
